src/updater.cpp

```diff
@@ -139,48 +139,44 @@ namespace appimage {
 
                     updateInformation = rawUpdateInformation.data();
                 } else if (appImageType == 2) {
-                    // check whether update information can be found inside the file by calling objdump
-
-                    // first of all, check whether there is an objdump binary next to the current one (probably
-                    // the bundled one in the AppImages of AppImageUpdate), otherwise use the system wide
-                    std::string objdump;
-
-                    {
-                        // TODO: replace this Linux specific solution with something platform independent
-                        std::vector<char> buffer(4096);
-                        readlink("/proc/self/exe", buffer.data(), buffer.size());
-
-                        auto pathToBinary = std::string(buffer.data());
-                        auto slashPos = pathToBinary.find_last_of('/');
-                        auto bundledObjdump = pathToBinary.substr(0, slashPos) + "/objdump";
-
-                        if (isFile(bundledObjdump)) {
-                            objdump = bundledObjdump;
-                        } else {
-                            objdump = "objdump";
+                   {
+                     unsigned offset = 0, length = 0;
+                     uint8_t * data;
+                     Elf64_Ehdr * elf64;
+                     Elf64_Shdr * shdr64;
+                     Elf32_Ehdr * elf32;
+                     Elf32_Shdr * shdr32;
+                     int i, fd = open(pathToAppImage.c_str(), O_RDONLY);
+                     data = (uint8_t * ) mmap(NULL, lseek(fd, 0, SEEK_END), PROT_READ, MAP_SHARED, fd, 0);
```

Comment:
This should be replaced by a less error prone C++ solution.
Bracket use breaks code style.

Reply:
What Can I replace mmap in C++11 ? 

In Qt I can simply do this to map a file safely...
```
QFile AppImage("Some.AppImage");
if(AppImage.open(QIODevice::ReadOnly)){
  uchar *Mapped = AppImage.map(0 , AppImage.size());
  uint8_t *data = (uint8_t*) Mapped;
  AppImage.unmap(Mapped);
  AppImage.close();
}
```

Any Idea's ?

Reply:
Sorry for not having been clear about this.

I don't like `mmap()` in general, as mapped files could, in theory, be altered in the background, hence changing the data in the map as well, which can cause some hard to debug bugs. Reading files using `std::ifstream` or even the C API (`read()` etc.) doesn't have such problems.

But as we're a) just reading from the file and b) the ELF API expects a char pointer, it is okay as-is... (I wouldn't have built it that way, but well...). So, sorry for being unclear, there is actually no issue here, I didn't read the rest of the code below.

Looking at the rest of the code, this seems safe to use. I just have a question: is the file unmapped at some point? I don't see a `munmap` call anywhere.

P.S.: My pointer-free alternative to using `mmap()` or `read()` or anything alike:

```cpp
#include <algorithm>
#include <fstream>
#include <string>

int main() {
    // ifstream doesn't provide any write facilities
    std::ifstream ifs("testfile");

    // checks if file could be opened
    if (ifs) {
        // initialize a new vector and resize it to the file size
        auto fileSize = ifs.tellg();
        std::vector<char> fileData(fileSize);

        // this is how you could seek to file's beginning
        // however, since we just opened the file, we're reading from the beginning already
        //ifs.clear();
        //ifs.seekg(0, std::ios::beg);

        // now, read in data
        ifs.read(fileData.data(), fileSize);

        // do something meaningful
        // ...

        return 0;
    }

    return 1;
}
```

Since data is a stack object, it will be deleted automatically once its scope ends. No cleanup required at all, no memory leaks, etc.. Pretty convenient to use, in my opinion.

Reply:
> Looking at the rest of the code, this seems safe to use. I just have a question: is the file unmapped at some point? I don't see a ```munmap``` call anywhere.

Yes , I forgot to unmap the mapped! 

Then I will be using **fstream** to find the filesize instead of the **C API**. And will use stack memory instead of heap for the **read** calls from **fstream**.

One thing that bothers me is that we have to open the file twice ( One for the stream and another for the **EFI API** ). (In Qt there is only one call to the file ).
For this , I found some compiler specific solutions on fstream using a file descriptor ( Can't use it because its gcc specific)

Thank You for your quick response! 👍

Reply:
> use stack memory

For the sake of completeness: I am pretty sure the `std::vector<...>` class does allocate its underlying arrays on the heap. 

Just keep using the C API in this case, if you like. Introducing an `fstream` just because we can doesn't make sense. Especially since we have a C file pointer anyway. As said before, it was my mistake, as I didn't fully understand what you're doing here at first.

> EFI

ELF *